### PR TITLE
Replace and remove feature gating class use with get environment type

### LINKED ISSUE #47694: Use WC_ENVIRONMENT_TYPE to check for development environment in WC blocks development mode features
**Issue Description:**

Right now we use `FeatureGating` to determine if we are in a development environment

As part of the goal of https://github.com/woocommerce/woocommerce/issues/43073 to remove `WOOCOMMERCE_BLOCKS_PHASE` and remove `FeatureGating` this task is to replace these 2 instances with the `WC_ENVIRONMENT_TYPE` / https://developer.wordpress.org/reference/functions/wp_get_environment_type/ API and this will allow us to get one step closer to removing `FeatureGating` altogether.

EDIT: We can't remove the FeatureGating class right away here due to still using the dev/prod env check, but we will be able to remove `WOOCOMMERCE_BLOCKS_PHASE` altogether.

### PR DESCRIPTION
**Pull request by louwie17**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This replaces the two instances left where we used the `FeatureGating` class and replaced it with the `wp_get_environment_type()` function.
The PR also deprecates the `FeatureGating` class.

Closes #47694, and https://github.com/woocommerce/woocommerce/issues/47695 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Build this branch and update a couple of the templates like checkout and cart, make sure everything works correctly ( without errors in the log )
2. Visit the checkout and cart pages on the front end and make sure everything works correctly ( without errors in the log )
3. Set the `WP_ENVIRONMENT_TYPE` constant to `local` or `development` in your `wp-config.php` file and make sure `SCRIPT_DEBUG` is set to `false`.
4. Check cart/checkout blocks load normally on the frontend.
5. To see the contents of the transient, when running a production build (and assuming you are using WP Local), open database > adminer and check the wp_options table for the option named _transient_woocommerce_blocks_asset_api_script_data. It should not exist, as it doesn't enable caching in `local` or `development` environments.
6. Set the `WP_ENVIRONMENT_TYPE` constant to `production` in your `wp-config.php` file and make sure `SCRIPT_DEBUG` is set to `false`.
7. Check cart/checkout blocks load normally on the frontend. 
8. Check the transient option again, it should contain JSON containing the list of assets/dependencies etc.
9. Install and enable this plugin from this Gist -> https://gist.github.com/louwie17/676721680b03762e88946c0270073f13 
[test-featuregating-plugin.zip](https://github.com/user-attachments/files/17734281/test-featuregating-plugin.zip)
10. Open the sites log file, it should show two debug messages:
```
The FeatureGating argument is deprecated since version 9.5. FeatureGating class is deprecated, please use wp_get_environment_type() instead.
```
and
```
Automattic\WooCommerce\Blocks\Domain\Services\FeatureGating::is_production_environment is deprecated since version 9.5.0! Use ::is_production_environment instead.
```

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/DeprecatedClassFacade.php [hunk 1]
```diff
@@ -58,7 +58,9 @@ class DeprecatedClassFacade {
 	 * Constructor.
 	 */
 	public function __construct() {
-		$this->instance = new static::$facade_over_classname();
+		if ( '' !== static::$facade_over_classname ) {
+			$this->instance = new static::$facade_over_classname();
```
Comment (louwie17) [inline plugins/woocommerce/src/Admin/DeprecatedClassFacade.php:62]: The `FeatureGating` class doesn't have a replacement class, so I updated this to allow for that.

This way it still catches the class initialization and any function calls and only shows the deprecated message.

File: plugins/woocommerce/src/Blocks/Assets/Api.php [hunk 2]
```diff
@@ -258,7 +258,7 @@ public function register_script( $handle, $relative_src, $dependencies = [], $ha
 		$script_data = $this->get_script_data( $relative_src, $dependencies );
 
 		if ( in_array( $handle, $script_data['dependencies'], true ) ) {
-			if ( $this->package->feature()->is_development_environment() ) {
+			if ( wp_get_environment_type() === 'development' ) {
```
Comment (louwie17) [inline plugins/woocommerce/src/Blocks/Assets/Api.php:261]: I am not sure how to test this condition, I did add a test case in the PR for the `disable_cache` option.

File: plugins/woocommerce/src/Blocks/Domain/Package.php [hunk 3]
```diff
@@ -47,12 +47,15 @@ class Package {
 	 *
 	 * @param string        $version        Version of the plugin.
 	 * @param string        $plugin_path    Path to the main plugin file.
-	 * @param FeatureGating $feature_gating Feature gating class instance.
+	 * @param FeatureGating $deprecated 	Deprecated Feature gating class.
```
Comment (louwie17) [inline plugins/woocommerce/src/Blocks/Domain/Package.php]: Given this is a public class, I kept the class construct signature the same, only added a default value for the `deprecated` param.

File: plugins/woocommerce/src/Blocks/Domain/Services/FeatureGating.php [hunk 4]
```diff
@@ -1,82 +1,29 @@
...
+	 * The version that this class was deprecated in.
 	 *
 	 * @var string
 	 */
-	private $environment;
-
-	const PRODUCTION_ENVIRONMENT  = 'production';
-	const DEVELOPMENT_ENVIRONMENT = 'development';
-	const TEST_ENVIRONMENT        = 'test';
+	protected static $deprecated_in_version = '9.5.0';
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/Domain/Services/FeatureGating.php]: Let's update the version. I think that the right one should be `9.6.0`.

File: plugins/woocommerce/src/Blocks/Domain/Package.php [hunk 5]
```diff
@@ -47,12 +47,15 @@ class Package {
...
 	 */
-	public function __construct( $version, $plugin_path, FeatureGating $feature_gating ) {
+	public function __construct( $version, $plugin_path, $deprecated = null ) {
+		if ( null !== $deprecated ) {
+			wc_deprecated_argument( 'FeatureGating', '9.5', 'FeatureGating class is deprecated, please use wp_get_environment_type() instead.' );
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/Domain/Package.php]: Let's update the version. I think that the right one should be `9.6.0`.

File: plugins/woocommerce/src/Blocks/Package.php [hunk 6]
```diff
@@ -66,9 +66,11 @@ public static function get_path() {
 	 * Returns an instance of the FeatureGating class.
 	 *
 	 * @return FeatureGating
+	 * @deprecated since 9.5.0, use wp_get_environment_type() instead.
 	 */
 	public static function feature() {
-		return self::get_package()->feature();
+		wc_deprecated_function( 'Package::feature', '9.5.0', 'wp_get_environment_type' );
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/Package.php]: Let's update the version. I think that the right one should be `9.6.0`.

**Review by gigitux — APPROVED** [review]
Sorry for the delay in reviewing this PR. Everything works as expected! Before merging, we just need to fix the WooCommerce version.

File: plugins/woocommerce/src/Blocks/Domain/Services/FeatureGating.php (hunk 4 above)
Comment (louwie17) [reply plugins/woocommerce/src/Blocks/Domain/Services/FeatureGating.php]: Done: https://github.com/woocommerce/woocommerce/pull/52768/commits/e5b60aec07e1857841d81ef54caf64b53e0d157e

File: plugins/woocommerce/src/Blocks/Domain/Package.php (hunk 5 above)
Comment (louwie17) [reply plugins/woocommerce/src/Blocks/Domain/Package.php]: Done: https://github.com/woocommerce/woocommerce/pull/52768/commits/e5b60aec07e1857841d81ef54caf64b53e0d157e

File: plugins/woocommerce/src/Blocks/Package.php (hunk 6 above)
Comment (louwie17) [reply plugins/woocommerce/src/Blocks/Package.php]: Done: https://github.com/woocommerce/woocommerce/pull/52768/commits/e5b60aec07e1857841d81ef54caf64b53e0d157e

**Comment by louwie17** [comment]
> Sorry for the delay in reviewing this PR. Everything works as expected! Before merging, we just need to fix the WooCommerce version.

Thanks, just updated the versions if you want to give it a quick glance again.

File: plugins/woocommerce/src/Blocks/Package.php [hunk 7]
```diff
@@ -66,9 +66,11 @@ public static function get_path() {
 	 * Returns an instance of the FeatureGating class.
 	 *
 	 * @return FeatureGating
+	 * @deprecated since 9.6.0, use wp_get_environment_type() instead.
 	 */
 	public static function feature() {
-		return self::get_package()->feature();
+		wc_deprecated_function( 'Package::feature', '9.6.0', 'wp_get_environment_type' );
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/Package.php]: Should it be `9.6`?

```suggestion
		wc_deprecated_function( 'Package::feature', '9.6', 'wp_get_environment_type' );
```

**Comment by gigitux** [comment]
> > Sorry for the delay in reviewing this PR. Everything works as expected! Before merging, we just need to fix the WooCommerce version.
> 
> Thanks, just updated the versions if you want to give it a quick glance again.

I added a comment! For the rest, LGTM!

File: plugins/woocommerce/src/Blocks/Package.php (hunk 7 above)
Comment (louwie17) [reply plugins/woocommerce/src/Blocks/Package.php]: Looks like we use both, but addressed here: https://github.com/woocommerce/woocommerce/pull/52768/commits/b0c26654e3467e0bb91fcf1eabc34b99d2bdbf03

**Pull request merged**